File: public/java/005-product-of-array-except-self.java
/*
 * LeetCode Problem #238: Product of Array Except Self
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/product-of-array-except-self/
 */
import java.util.*;

class Solution {

    // APPROACH 1: Brute Force | O(n²) time | O(1) space
    // EXPLAIN: For each index, multiply every other element together in a nested loop.
    // WHEN: Tiny arrays only; illustrates why the O(n) approaches are needed.
    public int[] productExceptSelf_BruteForce(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        for (int i = 0; i < n; i++) {
            int product = 1;
            for (int j = 0; j < n; j++) {
                if (j != i) product *= nums[j];
            }
            result[i] = product;
        }
        return result;
    }

    // APPROACH 2: Division | O(n) time | O(1) space
    // EXPLAIN: Compute total product, then divide by each element; zeros handled via counting.
    // WHEN: Shown for completeness only — the problem explicitly forbids using division.
    public int[] productExceptSelf_Division(int[] nums) {
        int n = nums.length;
        int zeroCount = 0;
        long product = 1;
        for (int x : nums) {
            if (x == 0) zeroCount++;
            else product *= x;
        }
        int[] result = new int[n];
        for (int i = 0; i < n; i++) {
            if (zeroCount > 1)       result[i] = 0;
            else if (zeroCount == 1) result[i] = (nums[i] == 0) ? (int) product : 0;
            else                     result[i] = (int) (product / nums[i]);
        }
        return result;
    }

    // APPROACH 3: Prefix + Suffix Arrays | O(n) time | O(n) space
    // EXPLAIN: Build a prefix product array and a suffix product array, then multiply them.
    // WHEN: Clear and readable; use when extra memory is acceptable.
    public int[] productExceptSelf_PrefixSuffix(int[] nums) {
        int n = nums.length;
        int[] prefix = new int[n];
        int[] suffix = new int[n];
        int[] result = new int[n];

        prefix[0] = 1;
        for (int i = 1; i < n; i++) {
            prefix[i] = prefix[i - 1] * nums[i - 1];
        }

        suffix[n - 1] = 1;
        for (int i = n - 2; i >= 0; i--) {
            suffix[i] = suffix[i + 1] * nums[i + 1];
        }

        for (int i = 0; i < n; i++) {
            result[i] = prefix[i] * suffix[i];
        }
        return result;
    }

    // APPROACH 4: Optimized Space | O(n) time | O(1) space
    // EXPLAIN: Use the output array for the prefix pass, then accumulate suffix in a running variable.
    // WHEN: Optimal — meets the problem's no-extra-space constraint.
    public int[] productExceptSelf_ConstantSpace(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];

        result[0] = 1;
        for (int i = 1; i < n; i++) {
            result[i] = result[i - 1] * nums[i - 1];
        }

        int suffixProduct = 1;
        for (int i = n - 1; i >= 0; i--) {
            result[i] *= suffixProduct;
            suffixProduct *= nums[i];
        }
        return result;
    }
}

// Made with Bob

File: public/python/005-product-of-array-except-self.py
# LeetCode Problem #238: Product of Array Except Self
# Difficulty: Medium
# Link: https://leetcode.com/problems/product-of-array-except-self/

# ─────────────────────────────────────────────
# APPROACH 1: Brute Force | O(n²) time | O(1) space
# EXPLAIN: For each index, multiply all other elements together in a nested loop.
# WHEN: Tiny arrays only; shows why the O(n) approaches exist.

def product_except_self_brute(nums: list[int]) -> list[int]:
    n = len(nums)
    result = []
    for i in range(n):
        product = 1
        for j in range(n):
            if j != i:
                product *= nums[j]
        result.append(product)
    return result


# ─────────────────────────────────────────────
# APPROACH 2: Division | O(n) time | O(1) space
# EXPLAIN: Compute the total product, then divide by each element; handle zeros specially.
# WHEN: Shown for completeness only — the problem statement explicitly forbids division.

def product_except_self_division(nums: list[int]) -> list[int]:
    n = len(nums)
    zero_count = nums.count(0)
    product = 1
    for x in nums:
        if x != 0:
            product *= x
    result = []
    for x in nums:
        if zero_count > 1:
            result.append(0)
        elif zero_count == 1:
            result.append(product if x == 0 else 0)
        else:
            result.append(product // x)
    return result


# ─────────────────────────────────────────────
# APPROACH 3: Prefix + Suffix Arrays | O(n) time | O(n) space
# EXPLAIN: Build a prefix-product array and a suffix-product array; answer[i] = prefix[i-1] * suffix[i+1].
# WHEN: When you want an explicit and readable O(n) solution using two passes.

def product_except_self_prefix_suffix(nums: list[int]) -> list[int]:
    n = len(nums)
    prefix  = [1] * n
    suffix  = [1] * n
    for i in range(1, n):
        prefix[i] = prefix[i - 1] * nums[i - 1]
    for i in range(n - 2, -1, -1):
        suffix[i] = suffix[i + 1] * nums[i + 1]
    return [prefix[i] * suffix[i] for i in range(n)]


# ─────────────────────────────────────────────
# APPROACH 4: Optimized Space | O(n) time | O(1) space
# EXPLAIN: Fill the output array with prefix products, then multiply in the suffix product on a second pass.
# WHEN: Optimal solution — constant extra space (the output array is not counted by convention).

def product_except_self_optimized(nums: list[int]) -> list[int]:
    n = len(nums)
    result = [1] * n
    # forward pass: result[i] holds product of all elements to the left of i
    for i in range(1, n):
        result[i] = result[i - 1] * nums[i - 1]
    # backward pass: multiply in the suffix product
    suffix = 1
    for i in range(n - 2, -1, -1):
        suffix *= nums[i + 1]
        result[i] *= suffix
    return result


# ─────────────────────────────────────────────
if __name__ == "__main__":
    cases = [
        ([1, 2, 3, 4],  [24, 12, 8, 6]),
        ([-1, 1, 0, -3, 3], [0, 0, 9, 0, 0]),
    ]
    for nums, expected in cases:
        assert product_except_self_brute(nums)           == expected
        assert product_except_self_prefix_suffix(nums)   == expected
        assert product_except_self_optimized(nums)       == expected
    # Division approach verified separately (integer-only, no zeros edge case here)
    assert product_except_self_division([1, 2, 3, 4]) == [24, 12, 8, 6]
    print("All tests passed.")

# Made with Bob
